### DISCUSSION TITLE
Gossipsub – subscriptions not propagated?

### DISCUSSION THREAD

**Discussion by Wiezzel** [libp2p/rust-libp2p#4393] [Q&A]
I'm running three nodes, let's call them `bootnode`, `subscriber`, and `publisher`. They use the following network behaviours: identify, kademlia, and gossipsub. Both `subscriber` and `publisher` connect to the `bootnode` and do a kademlia bootstrap.

```mermaid
graph TD;
    subscriber-->bootnode;
    publisher-->bootnode;
```

Then, `subscriber` subscribes to a gossipsub topic, and `publisher` starts publishing some messages on this topic. The problem is that messages are delivered only if `subscriber` **is started after** `publisher`. So it seems as if existing subscriptions are not propagated to a new node joining the network. Is this an expected behaviour or a bug? What could be a workaround for this?

**Comment by mxinden** [comment]
> The problem is that messages are delivered only if subscriber is started after publisher. 

This seems like a bug to me.

Is the `bootnode` subscribed to the topic?

> So it seems as if existing subscriptions are not propagated to a new node joining the network.

I don't think subscriptions are intended to be gossiped. In other words, I don't think subscription events go further than one hop.

> Is this an expected behaviour or a bug? What could be a workaround for this?

Are you able to create a small reproducer or even better write a small test that fails?

See https://github.com/libp2p/rust-libp2p/blob/master/protocols/gossipsub/tests/smoke.rs on how to write a gossipsub test.

**Reply by Wiezzel** [reply]
>Is the bootnode subscribed to the topic?

No, it is not.

I'll try to create a minimum reproducible example and share it.

**Reply by Wiezzel** [reply]
When I tried to build a minimal reproducible example, I couldn't reproduce the issue. So I got back to my original code and discovered that the problem doesn't occur if I wait 5s before starting `publisher`. Originally, `publisher` and `subscriber` connected to the `bootnode` roughly at the same time. Must be some kind of race condition. 🤔

**Reply by dariusc93** [reply]
can you run it with `RUST_LOG=libp2p_gossipsub=trace` and produce a log? Also could you provide any configurations, etc?

**Reply by Wiezzel** [reply]
@dariusc93 I'll try to prepare a minimal docker-compose setup with 3 nodes and share it.

**Reply by thomaseizinger** [reply]
Possibly related: https://github.com/libp2p/rust-libp2p/issues/3216.
